pkg/reconciler/pipelinerun/pipelinerun.go

```diff
@@ -566,12 +586,25 @@ func (c *Reconciler) reconcile(ctx context.Context, pr *v1beta1.PipelineRun) err
 		}
 	}
 
+	// candidateTasks contains a list of pipeline tasks that can be scheduled next
+	// This list of candidates is derived based on the successfully finished tasks
+	// A task is considered as candidate if its all parents have finished executing successfully
+	// candidateTasks are initialized with all root/s of DAG tasks
 	candidateTasks, err := dag.GetSchedulable(d, pipelineState.SuccessfulPipelineTaskNames()...)
 	if err != nil {
 		c.Logger.Errorf("Error getting potential next tasks for valid pipelinerun %s: %v", pr.Name, err)
 	}
 
+	// GetNextTasks returns a list of tasks which should be executed next and is derived based on candidateTasks
+	// GetNextTasks returns a list of candidates for which pipelineState does not have any taskRun or a list of
+	// failed tasks which haven't exhausted their retries
+	// Pipeline execution continues until GetNextTasks does not return any new pipeline task to execute
 	nextRprts := pipelineState.GetNextTasks(candidateTasks)
+	// GetFinalTasks returns a list of final tasks without any taskRun associated with it
+	// GetFinalTasks returns final tasks only when all DAG tasks have finished executing successfully or
+	// executing any one DAG task resulted in failure
+	nextRprts = append(nextRprts, pipelineState.GetFinalTasks(d, dfinally)...)
```

Comment:
im surprised this is being called even if nextRprts is non-0, b/c i thought that having values in nextRprts would indicate we are not ready to run finally tasks - would it make sense to check the length of nextRprts before calling this?

Reply:
I was surprised too, its tricky and we need to fix this.

Yes thats how it works in case of tasks getting executed and have finished successfully - nextRprts is non-0 means not ready to run finally tasks yet. But the same does not hold true in case of skipped root task (task with condition failure), that root skipped task is part of nextRprts always because being root is always part of candidates, and added to nextRprts since taskRun is nil (taskRun never gets created for such skipped task):

```
		if _, ok := candidateTasks[t.PipelineTask.Name]; ok && t.TaskRun == nil {
			tasks = append(tasks, t)
		}
```

https://github.com/tektoncd/pipeline/blob/58c836d55ef750cd023bae901fec51389a97f318/pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go#L189

Reply:
@bobcatfish I modified this code so that nextRprts results in 0 when its time to run finally tasks.

Updated `candidateTasks` retrieval such that a list of successful pipeline tasks and skipped tasks are used while calling `dag.GetSchedulable`:

```
		successOrSkippedTasks := append(pipelineState.SuccessfulPipelineTaskNames(), pipelineState.SkippedPipelineTaskNames(d)...)
	candidateTasks, err := dag.GetSchedulable(d, successOrSkippedTasks...)
```

This should fix `GetNextTasks` returning skipped tasks, please take a look 👍